File: 55/solution.py
#!/usr/bin/env python

def reverse_an_integer(n):
  """Reverse the order of digits in an integer"""
  return int(''.join(reversed(str(n))))

def is_palindrome(n):
  """Returns true if n is a palindromic number, false otherwise"""
  return reverse_an_integer(n) == n

def lychrel(n):
  """Returns true if n is probably a Lychrel number, false otherwise"""
  for i in xrange(50):
    n += reverse_an_integer(n)
    if is_palindrome(n):
      return False
  return True

def solution():
  """Returns the number of positive integers below 10k
     that are probably Lychrel numbers"""
  return len(filter(lychrel, xrange(1,10001)))

if __name__ == "__main__":
  print solution()

File: 55/solution.clj
;;; Project Euler 55

(defn reverse-integer [n]
  "reverse the order of digits in n"
  (BigInteger. (apply str (reverse (str n)))))

(defn palindrome? [n]
  "true if n is the same forward and reverse"
  (= n (reverse-integer n)))

(defn lychrel? [n]
  "true if n is probably a Lychrel number"
  (loop [i 50 x (+ n (reverse-integer n))]
    (if (palindrome? x)
      false
      (if (= i 0)
        true
        (recur (dec i) (+ x (reverse-integer x)))))))

(defn solution [n]
  "the number of positive integers < n that are probably Lychrel numbers"
  (count (filter lychrel? (range 1 n))) "\n")

(time (print (solution 10001)))